### PR TITLE
Fix compactions

### PR DESCRIPTION
**Pull request by fabxc**
@Gouthamve

### CODE REVIEW

File: db.go [hunk 1]
```diff
@@ -677,7 +692,7 @@ func (a *dbAppender) appenderAt(t int64) (*metaAppender, error) {
 	}
 
 	var hb headBlock
-	for _, h := range a.db.appendable() {
+	for _, h := range a.db.appendableHeads() {
```
Comment (Gouthamve) [inline db.go]: Not sure how this works, do this function get called every iteration?

**Comment by fabxc** [comment]
> Compacting only if headFullness > 0.5 seems dangerous. Maybe we should check both highTs and now()

So that was exactly the problem before. We only compacted then and in benchmarks, when our intermediate 30s compaction check didn't hit before a new block was started, we never compacted properly.

The main fix here is to always compact everything that's not the two most recent blocks. That's the old behavior. Now additionally, if we have a chance to catch a 50% full block, we compact this a bit earlier too to distribute our memory spikes a bit.

We cannot reference `now` as TSDB is wall clock agnostic. We could allow passing in a custom `now()` function. But I think what he have here works pretty well in practice and frees us from another config option.

File: db.go [hunk 2]
```diff
@@ -325,37 +325,52 @@ func headFullness(h headBlock) float64 {
...
+	for _, h := range db.heads[:len(db.heads)-2] {
+		if h.ActiveWriters() > 0 {
+			return r
 		}
+		r = append(r, h)
 	}
+	// Add the 2nd last head if the last head is more than 50% filled.
+	// Compacting it early allows us to free its memory before allocating
+	// more for the next block and thus reduces spikes.
+	if h2 := db.heads[len(db.heads)-2]; headFullness(h2) >= 0.5 && h2.ActiveWriters() == 0 {
```
Comment (fabxc) [inline db.go]: If the next value is 40, the next block will be started. Then the 2nd most-recent block immediately moves to position 3 and will be compacted as part of that.

File: db.go (hunk 1 above)
Comment (fabxc) [inline db.go]: This section of `appenderAt` will only be executed the first time we have to figure out the appropriate head block for timestamp `t`. In the general case, we return within the for loop.

**Comment by fabxc** [comment]
I've no clue why the tests seem to block forever in Travis.

**Comment by fabxc** [comment]
Locking issue fixed. The test was holding an open querier when trying to open a second one. The changes happened to cause timing that exposed a pre-existing issue.
The first querier didn't return its read lock (R1) because the second querier was waiting to acquire a read lock (R2). `reloadBlocks` tried to acquire a write lock (W1), which blocked (R2) but was waiting for (R1) to be released.

Also found two other existing issues around locking order.
Fixed all three and clarified docs around it.

**Pull request merged**